From f8f344fd1cde7a7892bdfadc35274cf28f420754 Mon Sep 17 00:00:00 2001
From: Ran Shalit <rshalit@habana.ai>
Date: Mon, 20 Jul 2020 10:20:42 +0300
Subject: [PATCH] [SW-15905] gaudi: security: fix revoke fit create

Fix revoke key create according for proper loading
by hl-fw-loader.

Change-Id: I24b28440bd460adda8139f045b1ded1eae2a7466
Signed-off-by: Ran Shalit <rshalit@habana.ai>
---

diff --git a/board/habanalabs/common/key_revoke.its b/board/habanalabs/common/key_revoke.its
index f8e177d..d82f5d7 100644
--- a/board/habanalabs/common/key_revoke.its
+++ b/board/habanalabs/common/key_revoke.its
@@ -10,14 +10,6 @@
 
 	images {
 		/* Signing a fit requires an image */
-		dummy {
-			description = "Habana dummy";
-			data = /incbin/("@DUMMY_BIN@");
-			compression = "none";
-			hash {
-				algo = "sha256";
-			};
-		};
 		pub-key {
 			description = "Public key in DTB format";
 			data = /incbin/("@KEY_DTB@");
@@ -32,15 +24,18 @@
 	configurations {
 		default = "fw-loader-conf";
 		fw-loader-conf {
-			dummy = "dummy";
+			pub-key = "pub-key";
 			load-bootloader;
 			/* load-type is FIT_LOAD_TYPE_REVOKE (2) */
 			load-type = <2>;
 			description = "Habanalabs key revoke";
+			/* fw-loader expects loadbales for fir verification */
+			loadables = "pub-key";
 			signature {
 				algo = "sha256,rsa3072";
 				key-name-hint = "hl_rsa3072";
-				sign-images = "dummy";
+				/* Signing fit requires at least one image */ 
+				sign-images = "pub-key";
 			};
 		};
 	};
diff --git a/board/habanalabs/gaudi/create-key-revoke.sh b/board/habanalabs/gaudi/create-key-revoke.sh
index baf58e6..7e2f0c7 100755
--- a/board/habanalabs/gaudi/create-key-revoke.sh
+++ b/board/habanalabs/gaudi/create-key-revoke.sh
@@ -9,12 +9,10 @@
 DUMMY_IMAGE=${IMAGE_DIRETORY}/${DUMMY_IMAGE_NAME}
 REVOKE_ITS_TEMPLATE=${BR2_EXTERNAL}/board/habanalabs/common/key_revoke.its
 REVOKE_ITS_FILE=${BR2_EXTERNAL}/board/habanalabs/common/${HL_DEV}-key_revoke.its
-REVOKE_ITB_FILE=${BR2_EXTERNAL}/board/habanalabs/common/${HL_DEV}-key_revoke.itb
+REVOKE_ITB_FILE=$IMAGE_DIRETORY/${HL_DEV}-key_revoke.itb
 KEYS_DIR=${BR2_EXTERNAL}/board/habanalabs/common/key/u-boot/
 
-dd if=/dev/urandom of=$DUMMY_IMAGE bs=4 count=1
 ITS_SED_PARAMS+="s#@KEY_DTB@#${KEYS_DIR}/hl_rsa3072.dtb#g;"
 ITS_SED_PARAMS+="s/@HL_DEV@/${HL_DEV}/g;"
-ITS_SED_PARAMS+="s#@DUMMY_BIN@#${DUMMY_IMAGE}#g;"
 sed ${ITS_SED_PARAMS} ${REVOKE_ITS_TEMPLATE} > ${REVOKE_ITS_FILE}
 $MKIMAGE_FILE -D "-W no-unit_address_vs_reg" -f ${REVOKE_ITS_FILE} -k ${KEYS_DIR} ${REVOKE_ITB_FILE}
diff --git a/scripts/make/gaudi/Makefile b/scripts/make/gaudi/Makefile
index ea62911..96ef8b2 100644
--- a/scripts/make/gaudi/Makefile
+++ b/scripts/make/gaudi/Makefile
@@ -75,6 +75,10 @@
 gaudi-key-revoke-create:
 	${BR2_EXTERNAL}/board/habanalabs/$(HL_DEV)/create-key-revoke.sh
 
+.PHONY: gaudi-key-revoke-install
+gaudi-key-revoke-install:
+	-sshpass -p$(HL_PASSWD) $(SCP) $(IMAGES_DIR)/${HL_DEV}-key_revoke.itb	root@$(HL_INSTALL_LAB_IP):/lib/firmware/habanalabs/$(HL_DEV)/
+
 .PHONY: gaudi-otp-create
 gaudi-otp-create:
 	$(HOST_MAKE_ENV) ${BR2_EXTERNAL}/board/habanalabs/$(HL_DEV)/create-otp-image.sh
